graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -48,6 +53,290 @@ std::pair<torch::Tensor, torch::Tensor> Sort(
+// is modified so that there is slack to be able to align as needed.
+template <typename indptr_t, typename indices_t>
+struct AlignmentFunc {
+  static_assert(GPU_CACHE_LINE_SIZE % sizeof(indices_t) == 0);
+  const indptr_t* in_degree;
+  const int64_t* perm;
+  int64_t num_nodes;
+  __host__ __device__ auto operator()(int64_t row) {
+    constexpr int num_elements = GPU_CACHE_LINE_SIZE / sizeof(indices_t);
+    return thrust::make_tuple(
+        in_degree[row],
+        // A single cache line has num_elements items, we add num_elements - 1
+        // to ensure there is enough slack to move forward or backward by
+        // num_elements - 1 items if the performed access is not aligned.
+        (indptr_t)(in_degree[perm ? perm[row % num_nodes] : row] + num_elements - 1));
+  }
+};
+
+template <typename indptr_t, typename indices_t>
+__global__ void _CSRRowWiseOneHopExtractorAlignedKernel(
+    const indptr_t hop_size, const int64_t num_nodes,
+    const indptr_t* const indptr, const indptr_t* const sub_indptr,
+    const indptr_t* const sub_indptr_aligned, const indices_t* const indices,
+    indices_t* const hop, const int64_t* const perm) {
+  indptr_t tx = static_cast<indptr_t>(blockIdx.x) * blockDim.x + threadIdx.x;
+  const int stride_x = gridDim.x * blockDim.x;
+
+  while (tx < hop_size) {
+    const auto rpos_ = cuda::UpperBound(sub_indptr_aligned, num_nodes, tx) - 1;
+    const auto rpos = perm ? perm[rpos_] : rpos_;
+    const auto out_row = sub_indptr[rpos];
+    const auto d = sub_indptr[rpos + 1] - out_row;
+    const int offset =
+        ((size_t)(indices + indptr[rpos] - sub_indptr_aligned[rpos_]) %
+         GPU_CACHE_LINE_SIZE) /
+        sizeof(indices_t);
+    const auto rofs = tx - sub_indptr_aligned[rpos_] - offset;
+    if (rofs >= 0 && rofs < d) {
+      const auto in_idx = indptr[rpos] + rofs;
+      assert((size_t)(indices + in_idx - tx) % GPU_CACHE_LINE_SIZE == 0);
+      const auto u = indices[in_idx];
+      hop[out_row + rofs] = u;
+    }
+    tx += stride_x;
+  }
+}
+
+// Given rows and indptr, computes:
+// inrow_indptr[i] = indptr[rows[i]];
+// in_deg[i] = indptr[rows[i] + 1] - indptr[rows[i]];
+template <typename indptr_t, typename nodes_t>
+struct DegreeFunc {
+  const nodes_t* rows;
+  const indptr_t* indptr;
+  indptr_t* in_deg;
+  indptr_t* inrow_indptr;
+  __host__ __device__ auto operator()(int64_t tIdx) {
+    const auto out_row = rows[tIdx];
+    const auto indptr_val = indptr[out_row];
+    const auto degree = indptr[out_row + 1] - indptr_val;
+    in_deg[tIdx] = degree;
+    inrow_indptr[tIdx] = indptr_val;
+  }
+};
+
+struct PairSum {
+  template <typename indptr_t>
+  __host__ __device__ auto operator()(
+      thrust::tuple<indptr_t, indptr_t> a,
+      thrust::tuple<indptr_t, indptr_t> b) {
+    return thrust::make_tuple(
+        thrust::get<0>(a) + thrust::get<0>(b),
+        thrust::get<1>(a) + thrust::get<1>(b));
+  };
+};
+
+template <typename indptr_t>
+auto ComputeDegree(
```

Comment:
As a utility function, can we return a pair of tensors?

Reply:
I allocate arrays as torch::Tensor only if they are going to be returned back to the python side. Otherwise, it is cheaper to allocate `unique_ptr`. If in the future this function is going to be used to return tensors back to python, we can change it. Currently, it is an internal implementation detail.

Reply:
Another advantage of `std::unique_ptr` are that we don't have to type the pointer type when using it. Simply writing `.get()` gives us a pointer with the right type, meaning the pointer type is specified during allocation only. Minimum risk of making coding mistakes or bugs.

Reply:
Considering this resolved, we can discuss it later for code style guidelines.

Reply:
I would prefer to use all tensor or all ptr if possible to keep consistency, unless there is a critical performance concern.

Reply:
My mental model is as follows: For any tensor for which `DISPATCH` has been applied, my rule of thumb is to switch to using pointers from that point on. However, for the `nodes` tensor, we have not used `DISPATCH` yet, it is dispatched inside the function. If both were made tensors, we would have nested dispatch for the same tensor, in particular, `indptr`. I feel that having too many dispatches will lead to too many variants of template parameters, possibly leading to exponentially longer compile times and code size.

Reply:
However, if a function proves to be useful elsewhere, we can refactor it out and use only `torch::tensor` as the datatype. In that case, we should never call that function inside any `DISPATCH` regions that overlap with the parameters of the function in discussion.

Reply:
SGTM, it is fine to have internal method to deviate slightly from the guidelines.